### ISSUE TITLE
Problems building with use_system_erfa=1 with 1.0.rc1

### ISSUE THREAD

**Issue by sergiopasra** [astropy/astropy#3381]
When building for Fedora, I pass (among others) the switch use_system_erfa=1 

For additional safety, I remove the erfa directory in cextern.  
Build worked in astropy < 1.0, use_system_erfa=1 means that the cextern/erfa directory is not ever accessed. 

But in 1.0rc1, I get the following error

```
+ /usr/bin/python2 setup.py build --offline
running build
running build_py
running pre_hook from astropy._erfa.setup_package for build_py command
pre_build_py_hook command hook build_py raised an exception: %s

Traceback (most recent call last):
  File "/home/spr/devel/fedora/python-astropy/astropy-1.0rc1/astropy_helpers/astropy_helpers/setup_helpers.py", line 398, in run_command_hooks
    hook_obj(cmd_obj)
  File "astropy/_erfa/setup_package.py", line 24, in pre_build_py_hook
    preprocess_source()
  File "astropy/_erfa/setup_package.py", line 72, in preprocess_source
    verbose=False)
  File "astropy/_erfa/erfa_generator.py", line 378, in main
    with open(erfahfn, "r") as f:
IOError: [Errno 2] No such file or directory: 'astropy/_erfa/../../cextern/erfa.h'
```
It seems to be trying to open a file in cextern, a file that doesn't exist (``cextern/erfa/erfa.h`` exists, but ``cextern/erfa.h`` doesn't)

**Comment by embray** [comment]
> For additional safety, I remove the erfa directory in cextern.

Not actually necessary, for what it's worth.    But it sounds like the `generate_erfa` script does not play correctly with `use_system_erfa`.

**Comment by mdboom** [comment]
The ERFA wrappers are generated directly from the ERFA source code (and not even the headers, but the .c file as well, see https://github.com/liberfa/erfa/issues/24).  Therefore it does need to be available, and it doesn't look for this elsewhere (since this can't be found through pkg-config or other automatic means, as far as I know).  All this needs to happen, even if linking against the system version.

However, the generated wrapper files are included with the tarball, so it you build from a tarball, this shouldn't be an issue -- it only matters if you are building from a git checkout.

**Comment by mdboom** [comment]
Correction: But I see it tries to list the `cextern/erfa` directory anyway.  That's not right...

**Comment by embray** [comment]
Right--nonetheless the interaction is buggy.

**Comment by mdboom** [comment]
How is it buggy?  It could provide a better error message for sure, but there isn't much else it can do if the source files are deleted.  The wrappers *can not* be generated from a system-installed erfa since it requires the .c files, not just the .h which are present in liberfa-devel.  If you have a tarball, though, this is a non-issue and (at least in my testing) it works fine with `--use-system-erfa` and `cextern/erfa` deleted.

**Comment by astrofrog** [comment]
Maybe I'm misunderstanding something, but @sergiopasra's original issue is about the RC candidate tarball, for which ``generate_erfa`` should not be trying to generate the wrappers since they should already be present and the source code shouldn't be used. So in that sense it's a bug since it's for a tarball distribution.

**Comment by sergiopasra** [comment]
@astrofrog Yes, this is about the RC tarball. 
What seems to happen is that ``preprocess_source`` triggers the regeneration of the interfaces every time.

In my case this test fails

```python
        erfa_mtime = max(os.path.getmtime(filename) for filename in SRC_FILES)
        gen_mtime = min(os.path.getmtime(filename) for filename in GEN_FILES)

        # If generated source is recent enough, don't update
        if gen_mtime > erfa_mtime:
            return
```
The mtimes are
```
[('astropy/_erfa/core.py.templ', 1422312625.0), ('astropy/_erfa/core.pyx.templ', 1422312625.0), ('astropy/_erfa/erfa_generator.py', 1422491051.7172613)]
[('astropy/_erfa/core.py', 1422491051.7572615), ('astropy/_erfa/core.pyx', 1422399952.0)]
```

The test files because ``erfa_generator`` is newer than one of the generated files.

**Comment by mdboom** [comment]
Ah, I see.  I had made my own tarball to test, which obviously wasn't thorough enough.  Maybe a `git clean` before generating the tarball would be enough to fix this...  Not sure how it would end up in that state in the first place.

**Comment by astrofrog** [comment]
If it does end up in an inconsistent state in term of the mtimes, we should probably just emit a warning instead of failing?

**Comment by eteq** [comment]
Huh, I cannot reproduce this.  If I do exactly what @sergiopasra  describes above and use the tarball to build, I get: 
```
('SRC', [('astropy/_erfa/core.py.templ', 1422312625.0), ('astropy/_erfa/core.pyx.templ', 1422312625.0), ('astropy/_erfa/erfa_generator.py', 1422312625.0)])
('GEN', [('astropy/_erfa/core.py', 1422399952.0), ('astropy/_erfa/core.pyx', 1422399952.0)])
```
That is, the generated files are all newer, so it proceeds without any problem.  @sergiopasra, is there any chance some part of the build process for fedora touches erfa_generator.py for some reason?  Or is it platform-dependent how well the timestamps are preserved?

In any event, @astrofrog raises a good point that for versions that are marked as release, we should just go ahead and use the generated versions anyway with an appropriate warning.  So in #3440 I implemented just that.

**Comment by sergiopasra** [comment]
@eteq  You are correct. If I check the mtimes in the tarball I get your numbers. 

But in the build process in fedora, after unpacking the tarball there is a 

```
find -name '*.py' | xargs sed -i '1s|^#!python|#!python2|'
```
sed seems to change the mtime regardless of if the substitution is actually made. I can try to do the same thing, but preserving the timestamp.

Anyway, I agree that #3440 is a good solution

**Comment by eteq** [comment]
Alright, this should be fixed now that #3440 is merged.  If not, feel free to reopen, @sergiopasra
